Inline free and sdallocx into operator delete

guangli-dai opened the pull request:
Inline free and sdallocx into operator delete for C++

guangli-dai committed:
Inline free and sdallocx into operator delete

interwq commented:
@guangli-dai please share the asm output of: 1) updated operator delete, 2) previous operator delete, and 3) regular free which the operator delete used to call.

guangli-dai commented:
1) The updated operator delete:
```
0000000000412460 <_ZdlPv>:
            (RTREE_CTX_NCACHE - 1));
  412460:       48 89 f8                mov    %rdi,%rax
        if (unlikely(rtree_ctx->cache[slot].leafkey != leafkey)) {
  412463:       48 c7 c2 f8 f4 ff ff    mov    $0xfffffffffffff4f8,%rdx
        return (key & mask);
  41246a:       48 89 f9                mov    %rdi,%rcx
        if (unlikely(rtree_ctx->cache[slot].leafkey != leafkey)) {
  41246d:       64 48 8b 34 25 00 00    mov    %fs:0x0,%rsi
  412474:       00 00 
            (RTREE_CTX_NCACHE - 1));
  412476:       48 c1 e8 1a             shr    $0x1a,%rax
        return (key & mask);
  41247a:       48 81 e1 00 00 00 c0    and    $0xffffffffc0000000,%rcx
        if (unlikely(rtree_ctx->cache[slot].leafkey != leafkey)) {
  412481:       25 f0 00 00 00          and    $0xf0,%eax
  412486:       48 01 f0                add    %rsi,%rax
  412489:       48 01 d0                add    %rdx,%rax
  41248c:       48 3b 88 b0 01 00 00    cmp    0x1b0(%rax),%rcx
  412493:       75 7b                   jne    412510 <_ZdlPv+0xb0>
        return ((key >> shiftbits) & mask);
  412495:       48 89 f9                mov    %rdi,%rcx
JEMALLOC_GENERATE_ATOMICS(void *, p, LG_SIZEOF_PTR)
  412498:       48 8b 80 b8 01 00 00    mov    0x1b8(%rax),%rax
  41249f:       48 c1 e9 0c             shr    $0xc,%rcx
  4124a3:       81 e1 ff ff 03 00       and    $0x3ffff,%ecx
  4124a9:       48 8b 0c c8             mov    (%rax,%rcx,8),%rcx
        contents.metadata.szind = bits >> LG_VADDR;
  4124ad:       48 89 c8                mov    %rcx,%rax
  4124b0:       48 c1 e8 30             shr    $0x30,%rax
                    &arena_emap_global, ptr, &alloc_ctx);

                /* Note: profiled objects will have alloc_ctx.slab set */
                if (unlikely(err || !alloc_ctx.slab ||
                    free_fastpath_nonfast_aligned(ptr,
                    /* check_prof */ false))) {
  4124b4:       83 e1 01                and    $0x1,%ecx
  4124b7:       74 57                   je     412510 <_ZdlPv+0xb0>
  4124b9:       89 c0                   mov    %eax,%eax

        uint64_t deallocated, threshold;
        te_free_fastpath_ctx(tsd, &deallocated, &threshold);
        size_t usize = sz_index2size(alloc_ctx.szind);
        uint64_t deallocated_after = deallocated + usize;
  4124bb:       48 8b 0c c5 80 a3 a9    mov    0xa9a380(,%rax,8),%rcx
  4124c2:       00 
  4124c3:       64 48 03 8a 48 03 00    add    %fs:0x348(%rdx),%rcx
  4124ca:       00 
         * 0) in a single branch.  Note that this handles the uninitialized case
         * as well (TSD init will be triggered on the non-fastpath).  Therefore
         * anything depends on a functional TSD (e.g. the alloc_ctx sanity check
         * below) needs to be after this branch.
         */
        if (unlikely(deallocated_after >= threshold)) {
  4124cb:       64 48 39 8a 50 03 00    cmp    %rcx,%fs:0x350(%rdx)
  4124d2:       00 
  4124d3:       76 3b                   jbe    412510 <_ZdlPv+0xb0>
        return n;
}
```

2) The assembly of the previous operator delete:
```
0000000000412460 <_ZdlPv>:

#endif  // __cpp_aligned_new

void
operator delete(void *ptr) noexcept {
        je_free(ptr);
  412460:       e9 9b 16 01 00          jmp    423b00 <__libc_free>
  412465:       66 66 2e 0f 1f 84 00    data16 cs nopw 0x0(%rax,%rax,1)
  41246c:       00 00 00 00

0000000000412470 <_ZdaPv>:
  412470:       e9 8b 16 01 00          jmp    423b00 <__libc_free>
  412475:       66 66 2e 0f 1f 84 00    data16 cs nopw 0x0(%rax,%rax,1)
  41247c:       00 00 00 00

0000000000412480 <_ZdlPvRKSt9nothrow_t>:
        je_free(ptr);
}

0000000000423b00 <__libc_free>:
        return (size_t)((key >> rtree_leaf_maskbits()) &
  423b00:       48 89 f8                mov    %rdi,%rax
        if (unlikely(rtree_ctx->cache[slot].leafkey != leafkey)) {
  423b03:       48 c7 c2 f8 f4 ff ff    mov    $0xfffffffffffff4f8,%rdx
        return (key & mask);
  423b0a:       48 89 f9                mov    %rdi,%rcx
        if (unlikely(rtree_ctx->cache[slot].leafkey != leafkey)) {
  423b0d:       64 48 8b 34 25 00 00    mov    %fs:0x0,%rsi
  423b14:       00 00
        return (size_t)((key >> rtree_leaf_maskbits()) &
  423b16:       48 c1 e8 1a             shr    $0x1a,%rax
        return (key & mask);
  423b1a:       48 81 e1 00 00 00 c0    and    $0xffffffffc0000000,%rcx
        if (unlikely(rtree_ctx->cache[slot].leafkey != leafkey)) {
  423b21:       25 f0 00 00 00          and    $0xf0,%eax
  423b26:       48 01 f0                add    %rsi,%rax
  423b29:       48 01 d0                add    %rdx,%rax
  423b2c:       48 3b 88 b0 01 00 00    cmp    0x1b0(%rax),%rcx
  423b33:       75 7b                   jne    423bb0 <__libc_free+0xb0>
        return ((key >> shiftbits) & mask);
  423b35:       49 89 f8                mov    %rdi,%r8
  423b38:       4c 8b 88 b8 01 00 00    mov    0x1b8(%rax),%r9
  423b3f:       49 c1 e8 0c             shr    $0xc,%r8
  423b43:       41 81 e0 ff ff 03 00    and    $0x3ffff,%r8d
  423b4a:       4f 8b 14 c1             mov    (%r9,%r8,8),%r10
        contents.metadata.szind = bits >> LG_VADDR;
  423b4e:       4d 89 d3                mov    %r10,%r11
  423b51:       49 c1 eb 30             shr    $0x30,%r11
                if (unlikely(err || !alloc_ctx.slab ||
  423b55:       41 83 e2 01             and    $0x1,%r10d
  423b59:       74 55                   je     423bb0 <__libc_free+0xb0>
  423b5b:       44 89 d8                mov    %r11d,%eax
        uint64_t deallocated_after = deallocated + usize;
  423b5e:       48 8b 0c c5 80 93 a9    mov    0xa99380(,%rax,8),%rcx
  423b65:       00
  423b66:       64 48 03 8a 48 03 00    add    %fs:0x348(%rdx),%rcx
  423b6d:       00
        if (unlikely(deallocated_after >= threshold)) {
  423b6e:       64 48 39 8a 50 03 00    cmp    %rcx,%fs:0x350(%rdx)
  423b75:       00
  423b76:       76 38                   jbe    423bb0 <__libc_free+0xb0>
        return ((uint16_t)(uintptr_t)bin->stack_head == bin->low_bits_full);
  423b78:       4c 8d 04 40             lea    (%rax,%rax,2),%r8
  423b7c:       4e 8d 0c c6             lea    (%rsi,%r8,8),%r9
  423b80:       49 01 d1                add    %rdx,%r9
  423b83:       49 8b b1 60 03 00 00    mov    0x360(%r9),%rsi
        if (unlikely(cache_bin_full(bin))) {
  423b8a:       66 41 39 b1 72 03 00    cmp    %si,0x372(%r9)
  423b91:       00
  423b92:       74 1c                   je     423bb0 <__libc_free+0xb0>
        bin->stack_head--;
  423b94:       4c 8d 56 f8             lea    -0x8(%rsi),%r10
  423b98:       4d 89 91 60 03 00 00    mov    %r10,0x360(%r9)
        *bin->stack_head = ptr;
  423b9f:       48 89 7e f8             mov    %rdi,-0x8(%rsi)
        *tsd_thread_deallocatedp_get(tsd) = deallocated_after;
  423ba3:       64 48 89 8a 48 03 00    mov    %rcx,%fs:0x348(%rdx)
  423baa:       00
}
  423bab:       c3                      ret
  423bac:       0f 1f 40 00             nopl   0x0(%rax)
                free_default(ptr);
  423bb0:       e9 3b f1 ff ff          jmp    422cf0 <je_free_default>
  423bb5:       66 66 2e 0f 1f 84 00    data16 cs nopw 0x0(%rax,%rax,1)
  423bbc:       00 00 00 00
```

3) The assembly of the regular free:
```
0000000000418010 <__libc_free>:
        return (size_t)((key >> rtree_leaf_maskbits()) &
  418010:       48 89 f8                mov    %rdi,%rax
        if (unlikely(rtree_ctx->cache[slot].leafkey != leafkey)) {
  418013:       48 c7 c2 58 f5 ff ff    mov    $0xfffffffffffff558,%rdx
        return (key & mask);
  41801a:       48 89 f9                mov    %rdi,%rcx
        if (unlikely(rtree_ctx->cache[slot].leafkey != leafkey)) {
  41801d:       64 48 8b 34 25 00 00    mov    %fs:0x0,%rsi
  418024:       00 00 
        return (size_t)((key >> rtree_leaf_maskbits()) &
  418026:       48 c1 e8 1a             shr    $0x1a,%rax
        return (key & mask);
  41802a:       48 81 e1 00 00 00 c0    and    $0xffffffffc0000000,%rcx
        if (unlikely(rtree_ctx->cache[slot].leafkey != leafkey)) {
  418031:       25 f0 00 00 00          and    $0xf0,%eax
  418036:       48 01 f0                add    %rsi,%rax
  418039:       48 01 d0                add    %rdx,%rax
  41803c:       48 3b 88 b0 01 00 00    cmp    0x1b0(%rax),%rcx
  418043:       75 7b                   jne    4180c0 <__libc_free+0xb0>
        return ((key >> shiftbits) & mask);
  418045:       49 89 f8                mov    %rdi,%r8
  418048:       4c 8b 88 b8 01 00 00    mov    0x1b8(%rax),%r9
  41804f:       49 c1 e8 0c             shr    $0xc,%r8
  418053:       41 81 e0 ff ff 03 00    and    $0x3ffff,%r8d
  41805a:       4f 8b 14 c1             mov    (%r9,%r8,8),%r10
        contents.metadata.szind = bits >> LG_VADDR;
  41805e:       4d 89 d3                mov    %r10,%r11
  418061:       49 c1 eb 30             shr    $0x30,%r11
                if (unlikely(err || !alloc_ctx.slab ||
  418065:       41 83 e2 01             and    $0x1,%r10d
  418069:       74 55                   je     4180c0 <__libc_free+0xb0>
  41806b:       44 89 d8                mov    %r11d,%eax
        uint64_t deallocated_after = deallocated + usize;
  41806e:       48 8b 0c c5 c0 57 98    mov    0x9857c0(,%rax,8),%rcx
  418075:       00 
  418076:       64 48 03 8a 48 03 00    add    %fs:0x348(%rdx),%rcx
  41807d:       00 
        if (unlikely(deallocated_after >= threshold)) {
  41807e:       64 48 39 8a 50 03 00    cmp    %rcx,%fs:0x350(%rdx)
  418085:       00 
        if (unlikely(deallocated_after >= threshold)) {
  41807e:       64 48 39 8a 50 03 00    cmp    %rcx,%fs:0x350(%rdx)
  418085:       00 
  418086:       76 38                   jbe    4180c0 <__libc_free+0xb0>
        return ((uint16_t)(uintptr_t)bin->stack_head == bin->low_bits_full);
  418088:       4c 8d 04 40             lea    (%rax,%rax,2),%r8
  41808c:       4e 8d 0c c6             lea    (%rsi,%r8,8),%r9
  418090:       49 01 d1                add    %rdx,%r9
  418093:       49 8b b1 60 03 00 00    mov    0x360(%r9),%rsi
        if (unlikely(cache_bin_full(bin))) {
  41809a:       66 41 39 b1 72 03 00    cmp    %si,0x372(%r9)
  4180a1:       00 
  4180a2:       74 1c                   je     4180c0 <__libc_free+0xb0>
        bin->stack_head--;
  4180a4:       4c 8d 56 f8             lea    -0x8(%rsi),%r10
  4180a8:       4d 89 91 60 03 00 00    mov    %r10,0x360(%r9)
        *bin->stack_head = ptr;
  4180af:       48 89 7e f8             mov    %rdi,-0x8(%rsi)
        *tsd_thread_deallocatedp_get(tsd) = deallocated_after;
  4180b3:       64 48 89 8a 48 03 00    mov    %rcx,%fs:0x348(%rdx)
  4180ba:       00 
}
  4180bb:       c3                      ret
  4180bc:       0f 1f 40 00             nopl   0x0(%rax)
                free_default(ptr);
  4180c0:       e9 2b f6 ff ff          jmp    4176f0 <je_free_default>
  4180c5:       66 66 2e 0f 1f 84 00    data16 cs nopw 0x0(%rax,%rax,1)
  4180cc:       00 00 00 00 
```

guangli-dai commented on include/jemalloc/internal/jemalloc_internal_inlines_d.h:
```diff
@@ -0,0 +1,220 @@
+#ifndef JEMALLOC_INTERNAL_INLINES_D_H
```
Since moving these functions around won't affect the interdependency, can they be put in other jemalloc_internal_inlines_*.h?

guangli-dai replied:
Cannot move them elsewhere because of the interdependency with prof_inlines.h

interwq replied:
Is `prof_sample_aligned` the only part needed from `prof_inlines.h`? If so let's take that out of the prof header, and probably rename to `nonfast_aligned` or `special_aligned`. Then you don't have to add yet another header?

guangli-dai replied:
Yes, since `prof_sample_aligned` is merely one line. I copied it and named it as `nonfast_aligned`. With this I can put everything in jemalloc_internal_inlines_c.h instead of creating a new file. I kept the original function in prof_inlines.h since it is needed there by other functions there as well.

guangli-dai committed:
Benchmark operator delete

Added the microbenchmark for operator delete.
Also modified bench.h so that it can be used in C++.

guangli-dai committed:
Update the ratio display in benchmark

In bench.h, specify the ratio as the time consumption ratio and
modify the display of the ratio.

guangli-dai commented:
Local benchmark results with inline:
```
$ sudo nice -19 numactl -C 4 sudo -u $USER sh test/test.sh test/stress/cpp/microbench
=== test/stress/cpp/microbench ===
100000000 iterations, new_size_delete=787018us (7.870 ns/iter), malloc_sdallocx=852019us (8.520 ns/iter), time consumption ratio=0.924:1
test_sized_delete_vs_sdallocx (non-reentrant): pass
100000000 iterations, malloc_free=1342031us (13.420 ns/iter), new_delete=1297030us (12.970 ns/iter), time consumption ratio=1.035:1
test_free_vs_delete (non-reentrant): pass
100000000 iterations, malloc_free_array=1311030us (13.110 ns/iter), delete_array=1289029us (12.890 ns/iter), time consumption ratio=1.017:1
test_free_array_vs_delete_array (non-reentrant): pass
--- pass: 3/3, skip: 0/3, fail: 0/3 ---
```

Local benchmark results without inline:
```
$ sudo nice -19 numactl -C 4 sudo -u $USER sh test/test.sh test/stress/cpp/microbench
=== test/stress/cpp/microbench ===
100000000 iterations, new_size_delete=938021us (9.380 ns/iter), malloc_sdallocx=863020us (8.630 ns/iter), time consumption ratio=1.087:1
test_sized_delete_vs_sdallocx (non-reentrant): pass
100000000 iterations, malloc_free=1334030us (13.340 ns/iter), new_delete=1225028us (12.250 ns/iter), time consumption ratio=1.089:1
test_free_vs_delete (non-reentrant): pass
100000000 iterations, malloc_free_array=1287029us (12.870 ns/iter), delete_array=1270029us (12.700 ns/iter), time consumption ratio=1.013:1
test_free_array_vs_delete_array (non-reentrant): pass
--- pass: 3/3, skip: 0/3, fail: 0/3 ---
```

Operator delete and operator delete[] do not show an obvious improvement here as expected. The reason could be that the rtree lookup is blocking the pipeline and make the save of the extra jump minor.

Sized delete does show a great decrease of time due to the inlining for around 15%.

interwq commented on test/integration/cpp/microbench.cpp:
```diff
@@ -0,0 +1,68 @@
+
+int
+main() {
+	return test(
```
For benchmarks, no need to test different reentrancy cases (i.e right now each test is done 3 times). You can use ` test_no_reentrancy()` here.

guangli-dai replied:
Modified

interwq commented:
For the local benchmark results, can you please also obtain:
1) compare the cost of delete w/ and w/o the inline commit. (i.e. run the benchmark w/ and w/o the inline change)
2) an estimation of % win, based on perf annotate -- meaning to quantify the cost of the instructions being optimized away; 
3) an estimation of % win, based on the percentage of operator delete (from of `perf report`), before and after -- e.g. if the total cost % of `delete` drops from 50% to 45%, it would be a roughly 10% reduction).

Let me know if any questions.

interwq commented on test/integration/cpp/microbench.cpp:
```diff
@@ -0,0 +1,68 @@
+
+TEST_BEGIN(test_delete_vs_sized_delete) {
+	compare_funcs(10*1000*1000, 100*1000*1000, "delete", 
+	    (void *)new_delete, "sized_delete", (void *)new_sized_delete);
```
Maybe compare `sdallocx_noflags` with sized delete here? Since we already measured the delete cost in the free_vs_delete case.

guangli-dai replied:
Modified. Also use a macro to make the test compatible on machines that do not allow sized delete

interwq commented:
You will also want to reorder the commits -- the inline change should be the last, so that you can run the benchmark w/ and w/o that. You can do the reordering through `git rebase -i dev`.

guangli-dai commented:
1. The assembly of the inline sized delete:
```
0000000000412760 <_ZdlPvm>:
#if __cpp_sized_deallocation >= 201309

JEMALLOC_ALWAYS_INLINE
void
sizedDeleteImpl(void* ptr, std::size_t size) noexcept {
        if (unlikely(ptr == nullptr)) {
  412760:       48 85 ff                test   %rdi,%rdi
  412763:       0f 84 97 00 00 00       je     412800 <_ZdlPvm+0xa0>
                    /* check_prof */ true))) {
  412769:       48 81 fe 00 10 00 00    cmp    $0x1000,%rsi
  412770:       77 7e                   ja     4127f0 <_ZdlPvm+0x90>
                if (prof_sample_aligned(ptr)) {
  412772:       f7 c7 ff 0f 00 00       test   $0xfff,%edi
  412778:       74 76                   je     4127f0 <_ZdlPvm+0x90>
        return sz_size2index_tab[(size + (ZU(1) << SC_LG_TINY_MIN) - 1)
  41277a:       48 8d 46 07             lea    0x7(%rsi),%rax
}

JEMALLOC_ALWAYS_INLINE void
te_free_fastpath_ctx(tsd_t *tsd, uint64_t *deallocated, uint64_t *threshold) {
        /* Unsafe getters since this may happen before tsd_init. */
        *deallocated = *tsd_thread_deallocatedp_get_unsafe(tsd);
  41277e:       48 c7 c1 f8 f4 ff ff    mov    $0xfffffffffffff4f8,%rcx
            >> SC_LG_TINY_MIN];
  412785:       48 c1 e8 03             shr    $0x3,%rax
        return sz_index2size_tab[index];
  412789:       0f b6 80 00 ab a9 00    movzbl 0xa9ab00(%rax),%eax
        uint64_t deallocated_after = deallocated + usize;
  412790:       48 8b 14 c5 80 a3 a9    mov    0xa9a380(,%rax,8),%rdx
  412797:       00
  412798:       64 48 03 91 48 03 00    add    %fs:0x348(%rcx),%rdx
  41279f:       00
        if (unlikely(deallocated_after >= threshold)) {
  4127a0:       64 48 39 91 50 03 00    cmp    %rdx,%fs:0x350(%rcx)
  4127a7:       00
  4127a8:       76 46                   jbe    4127f0 <_ZdlPvm+0x90>
        return ((uint16_t)(uintptr_t)bin->stack_head == bin->low_bits_full);
  4127aa:       4c 8d 04 40             lea    (%rax,%rax,2),%r8
  4127ae:       64 48 8b 04 25 00 00    mov    %fs:0x0,%rax
  4127b5:       00 00
  4127b7:       4a 8d 04 c0             lea    (%rax,%r8,8),%rax
  4127bb:       48 01 c8                add    %rcx,%rax
  4127be:       4c 8b 80 60 03 00 00    mov    0x360(%rax),%r8
        if (unlikely(cache_bin_full(bin))) {
  4127c5:       66 44 39 80 72 03 00    cmp    %r8w,0x372(%rax)
  4127cc:       00
  4127cd:       74 21                   je     4127f0 <_ZdlPvm+0x90>
        bin->stack_head--;
  4127cf:       49 8d 70 f8             lea    -0x8(%r8),%rsi
  4127d3:       48 89 b0 60 03 00 00    mov    %rsi,0x360(%rax)
        *bin->stack_head = ptr;
  4127da:       49 89 78 f8             mov    %rdi,-0x8(%r8)
        *tsd_thread_deallocatedp_get(tsd) = deallocated_after;
  4127de:       64 48 89 91 48 03 00    mov    %rdx,%fs:0x348(%rcx)
  4127e5:       00
        return true;
  4127e6:       c3                      ret
  4127e7:       66 0f 1f 84 00 00 00    nopw   0x0(%rax,%rax,1)
  4127ee:       00 00
                sdallocx_default(ptr, size, 0);
  4127f0:       31 d2                   xor    %edx,%edx
  4127f2:       e9 89 78 01 00          jmp    42a080 <je_sdallocx_default>
  4127f7:       66 0f 1f 84 00 00 00    nopw   0x0(%rax,%rax,1)
  4127fe:       00 00
}
```

2. The assembly of the previous sized delete
```
00000000004124a0 <_ZdlPvm>:
#if __cpp_sized_deallocation >= 201309

JEMALLOC_ALWAYS_INLINE
void
sizedDeleteImpl(void* ptr, std::size_t size) noexcept {
        if (unlikely(ptr == nullptr)) {
  4124a0:       48 85 ff                test   %rdi,%rdi
  4124a3:       74 13                   je     4124b8 <_ZdlPvm+0x18>
        }
        je_sdallocx_noflags(ptr, size);
}

void
operator delete(void *ptr, std::size_t size) noexcept {
  4124a5:       48 83 ec 08             sub    $0x8,%rsp
        je_sdallocx_noflags(ptr, size);
  4124a9:       e8 42 85 01 00          call   42a9f0 <je_je_sdallocx_noflags>
        sizedDeleteImpl(ptr, size);
}
  4124ae:       48 83 c4 08             add    $0x8,%rsp
  4124b2:       c3                      ret
  4124b3:       0f 1f 44 00 00          nopl   0x0(%rax,%rax,1)
  4124b8:       c3                      ret
  4124b9:       0f 1f 80 00 00 00 00    nopl   0x0(%rax)

000000000042a9f0 <je_je_sdallocx_noflags>:
                if (unlikely(size > SC_LOOKUP_MAXCLASS ||
  42a9f0:       48 81 fe 00 10 00 00    cmp    $0x1000,%rsi
  42a9f7:       77 77                   ja     42aa70 <je_je_sdallocx_noflags+0x80>
                if (prof_sample_aligned(ptr)) {
  42a9f9:       f7 c7 ff 0f 00 00       test   $0xfff,%edi
  42a9ff:       74 6f                   je     42aa70 <je_je_sdallocx_noflags+0x80>
        return sz_size2index_tab[(size + (ZU(1) << SC_LG_TINY_MIN) - 1)
  42aa01:       48 8d 46 07             lea    0x7(%rsi),%rax
  42aa05:       48 c7 c1 f8 f4 ff ff    mov    $0xfffffffffffff4f8,%rcx
            >> SC_LG_TINY_MIN];
  42aa0c:       48 c1 e8 03             shr    $0x3,%rax
        return sz_index2size_tab[index];
  42aa10:       44 0f b6 80 00 9b a9    movzbl 0xa99b00(%rax),%r8d
  42aa17:       00
        uint64_t deallocated_after = deallocated + usize;
  42aa18:       4a 8b 14 c5 80 93 a9    mov    0xa99380(,%r8,8),%rdx
  42aa1f:       00
  42aa20:       64 48 03 91 48 03 00    add    %fs:0x348(%rcx),%rdx
  42aa27:       00
        if (unlikely(deallocated_after >= threshold)) {
  42aa28:       64 48 39 91 50 03 00    cmp    %rdx,%fs:0x350(%rcx)
  42aa2f:       00
  42aa30:       76 3e                   jbe    42aa70 <je_je_sdallocx_noflags+0x80>
        return ((uint16_t)(uintptr_t)bin->stack_head == bin->low_bits_full);
  42aa32:       64 4c 8b 14 25 00 00    mov    %fs:0x0,%r10
  42aa39:       00 00
  42aa3b:       4f 8d 0c 40             lea    (%r8,%r8,2),%r9
  42aa3f:       4f 8d 1c ca             lea    (%r10,%r9,8),%r11
  42aa43:       49 01 cb                add    %rcx,%r11
  42aa46:       49 8b 83 60 03 00 00    mov    0x360(%r11),%rax
        if (unlikely(cache_bin_full(bin))) {
  42aa4d:       66 41 39 83 72 03 00    cmp    %ax,0x372(%r11)
  42aa54:       00
  42aa55:       74 19                   je     42aa70 <je_je_sdallocx_noflags+0x80>
        bin->stack_head--;
  42aa57:       48 8d 70 f8             lea    -0x8(%rax),%rsi
  42aa5b:       49 89 b3 60 03 00 00    mov    %rsi,0x360(%r11)
        *bin->stack_head = ptr;
  42aa62:       48 89 78 f8             mov    %rdi,-0x8(%rax)
        *tsd_thread_deallocatedp_get(tsd) = deallocated_after;
  42aa66:       64 48 89 91 48 03 00    mov    %rdx,%fs:0x348(%rcx)
  42aa6d:       00
}
  42aa6e:       c3                      ret
  42aa6f:       90                      nop
                sdallocx_default(ptr, size, 0);
  42aa70:       31 d2                   xor    %edx,%edx
  42aa72:       e9 19 ee ff ff          jmp    429890 <je_sdallocx_default>
  42aa77:       66 0f 1f 84 00 00 00    nopw   0x0(%rax,%rax,1)
  42aa7e:       00 00
```
3. The assembly of the sdallocx 
```
0000000000428580 <sdallocx>:
                if (unlikely(size > SC_LOOKUP_MAXCLASS ||
  428580:       48 81 fe 00 10 00 00    cmp    $0x1000,%rsi
  428587:       77 7f                   ja     428608 <sdallocx+0x88>
  428589:       85 d2                   test   %edx,%edx
  42858b:       75 7b                   jne    428608 <sdallocx+0x88>
                if (prof_sample_aligned(ptr)) {
  42858d:       f7 c7 ff 0f 00 00       test   $0xfff,%edi
  428593:       74 73                   je     428608 <sdallocx+0x88>
        return sz_size2index_tab[(size + (ZU(1) << SC_LG_TINY_MIN) - 1)
  428595:       48 8d 46 07             lea    0x7(%rsi),%rax
        *deallocated = *tsd_thread_deallocatedp_get_unsafe(tsd);
  428599:       49 c7 c0 50 f5 ff ff    mov    $0xfffffffffffff550,%r8
            >> SC_LG_TINY_MIN];
  4285a0:       48 c1 e8 03             shr    $0x3,%rax
        return sz_index2size_tab[index];
  4285a4:       44 0f b6 88 00 95 9d    movzbl 0x9d9500(%rax),%r9d
  4285ab:       00
        uint64_t deallocated_after = deallocated + usize;
  4285ac:       4a 8b 0c cd 40 97 9d    mov    0x9d9740(,%r9,8),%rcx
  4285b3:       00
  4285b4:       64 49 03 88 48 03 00    add    %fs:0x348(%r8),%rcx
  4285bb:       00
        if (unlikely(deallocated_after >= threshold)) {
  4285bc:       64 49 39 88 50 03 00    cmp    %rcx,%fs:0x350(%r8)
  4285c3:       00
  4285c4:       76 42                   jbe    428608 <sdallocx+0x88>
        return ((uint16_t)(uintptr_t)bin->stack_head == bin->low_bits_full);
  4285c6:       64 4c 8b 1c 25 00 00    mov    %fs:0x0,%r11
  4285cd:       00 00
  4285cf:       4f 8d 14 49             lea    (%r9,%r9,2),%r10
  4285d3:       4b 8d 04 d3             lea    (%r11,%r10,8),%rax
  4285d7:       4c 01 c0                add    %r8,%rax
  4285da:       4c 8b 88 60 03 00 00    mov    0x360(%rax),%r9
        if (unlikely(cache_bin_full(bin))) {
  4285e1:       66 44 39 88 72 03 00    cmp    %r9w,0x372(%rax)
  4285e8:       00
  4285e9:       74 1d                   je     428608 <sdallocx+0x88>
        bin->stack_head--;
  4285eb:       49 8d 51 f8             lea    -0x8(%r9),%rdx
  4285ef:       48 89 90 60 03 00 00    mov    %rdx,0x360(%rax)
        *bin->stack_head = ptr;
  4285f6:       49 89 79 f8             mov    %rdi,-0x8(%r9)
        *tsd_thread_deallocatedp_get(tsd) = deallocated_after;
  4285fa:       64 49 89 88 48 03 00    mov    %rcx,%fs:0x348(%r8)
  428601:       00
}
  428602:       c3                      ret
  428603:       0f 1f 44 00 00          nopl   0x0(%rax,%rax,1)
                sdallocx_default(ptr, size, flags);
  428608:       e9 a3 ee ff ff          jmp    4274b0 <je_sdallocx_default>
  42860d:       0f 1f 00                nopl   (%rax)
```

guangli-dai commented:
Comparison on the percentage:
1. Before inlining:
```
Overhead  Command   Shared Object          Symbol
  46.99%  test_cpp  test_cpp               [.] free
  ...
   4.07%  test_cpp  test_cpp               [.] operator delete
  ...
```
The total percentage for the delete operation is 51.06%.

2. After inlining:
```
Overhead  Command   Shared Object                                  Symbol
  50.54%  test_cpp  test_cpp                                       [.] operator delete
  ...
```
"free" is now inlined and in total delete occupies 50.54%
So the cycle reduction is merely around 0.5% for delete.

3. Sized delete without inlining:
```
Overhead  Command         Shared Object        Symbol
  34.91%  test_sized_cpp  test_sized_cpp       [.] je_je_sdallocx_noflags
   ...
   9.36%  test_sized_cpp  test_sized_cpp       [.] operator delete
   ...
```
4. Sized delete with inlining:
```
Overhead  Command         Shared Object          Symbol
  39.47%  test_sized_cpp  test_sized_cpp         [.] operator delete
```
"je_je_sdallocx_noflags" is gone, which saves around 5% of time.

interwq commented:
> "je_je_sdallocx_noflags" is gone, which saves around 5% of time.

Since we are focusing on the sdallocx itself, the reduction should be more like: 1 - 39.47% / (9.36%+34.91%) = 10.8%. Similarly the non-sized delete change is ~1%. 

It looks you also updated the benchmark output? So the meaningful change from that is the new-sized-delete improved from 9.38ns to 7.87ns?

guangli-dai commented:
> It looks you also updated the benchmark output? So the meaningful change from that is the new-sized-delete improved from 9.38ns to 7.87ns?

Yes. They have been updated. The rough range of the improvement is around 5-15% based on the three different types of estimation.

interwq commented on Makefile.in:
```diff
@@ -362,9 +367,10 @@ TESTS_INTEGRATION_OBJS := $(TESTS_INTEGRATION:$(srcroot)%.c=$(objroot)%.$(O))
 TESTS_OBJS := $(TESTS_UNIT_OBJS) $(TESTS_INTEGRATION_OBJS) $(TESTS_ANALYZE_OBJS) \
-	$(TESTS_STRESS_OBJS)
-TESTS_CPP_OBJS := $(TESTS_INTEGRATION_CPP_OBJS)
+	$(TESTS_STRESS_OBJS) 
```
nit: unnecessary white space

guangli-dai replied:
removed

interwq commented on test/include/test/bench.h:
```diff
@@ -23,7 +23,7 @@ fmt_nsecs(uint64_t usec, uint64_t iters, char *buf) {
 	uint64_t intpart = nsecs_per_iter1000 / 1000;
 	uint64_t fracpart = nsecs_per_iter1000 % 1000;
-	malloc_snprintf(buf, FMT_NSECS_BUF_SIZE, "%"FMTu64".%03"FMTu64, intpart,
+	malloc_snprintf(buf, FMT_NSECS_BUF_SIZE, "%" FMTu64 ".%03" FMTu64, intpart,
```
nit: We don't really need the white space here. This part is the macro for the `%u` etc. in the printf format and we can put the strings together.

guangli-dai replied:
True, the spaces are not mandatory. Yet it will report a warning when compiling with c++11: `test/include/test/bench.h:55:16: warning: invalid suffix on literal; C++11 requires a space between literal and string macro [-Wliteral-suffix]
  malloc_printf("%"FMTu64" iterations, %s=%" FMTu64 "us (%s ns/iter), "` 
Thus, I modified it for better compatibility just in case.

interwq replied:
Sounds good. Let's certain avoid any warnings here.

interwq commented on test/stress/cpp/microbench.cpp:
```diff
@@ -0,0 +1,84 @@
+}
+TEST_END
+
+#if __cpp_sized_deallocation >= 201309
```
From this point below there are 3 `#if`s. We can merge them into one:
```
TEST_BEGIN(test_sized_delete_vs_sdallocx) {
#if __cpp_sized_deallocation >= 201309
    compare_funcs(...)
#else
    malloc_printf(...);
#endif
}
```

guangli-dai replied:
Merged

interwq commented on test/stress/cpp/microbench.cpp:
```diff
@@ -0,0 +1,83 @@
+static void
+malloc_free(void) {
+	void *p = malloc(1);
+	expect_ptr_not_null(p, "Unexpected new failure");
```
nit: error msg should say malloc instead of new (similar below in malloc_free_array)

interwq commented on include/jemalloc/internal/jemalloc_internal_inlines_c.h:
```diff
@@ -7,6 +7,12 @@
+
+#define TCACHE_IND_NONE ((unsigned)-1)
+#define TCACHE_IND_AUTOMATIC ((unsigned)-2)
+#define ARENA_IND_AUTOMATIC ((unsigned)-1)
```
We should remove the original definition in `jemalloc.c`:
https://github.com/jemalloc/jemalloc/blob/481bbfc9906e7744716677edd49d0d6c22556a1a/src/jemalloc.c#L2281

guangli-dai replied:
removed.

interwq commented on include/jemalloc/internal/jemalloc_internal_inlines_c.h:
```diff
@@ -337,4 +343,217 @@ imalloc_fastpath(size_t size, void *(fallback_alloc)(size_t)) {
 }
 
+JEMALLOC_ALWAYS_INLINE bool
+nonfast_aligned(const void *ptr) {
```
I just realized that we already have `free_fastpath_nonfast_aligned`. In that case this isn't a good naming choice. Since we are checking for different alignments in `free_fastpath_nonfast_aligned`, we should still call out which part this alignment is for. Let's still call it `prof_sample_aligned`.

We definitely don't want to duplicate the code though; even if it's only 1 line, it's part of the core logic (and would cause super confusing bugs the two copies are out of sync). Can we remove the one in `prof_inlines.h` and only keep this one (maybe even include this header in prof_inlines.h)?

guangli-dai replied:
I think we can. The function is used in both prof_inlines.h and jemalloc.c and they both can include this header. It seems my last comment asking this was not sent out. I wanted to ask about the duplicate. I agree that we should not duplicate, but if this is for prof purposes and start with a prefix `prof_`, will it be a bit weird to appear in this file and this file only?

interwq replied:
It's indeed not ideal. Maybe taking a step back and look into how the current header dependency is set up, and see if it's possible to keep it prof_inlines.h?

Even if we can't, it won't be too bad though, given that we also have other inline utilities included (e.g. the `tcache_` ones). Maintaining the DAG does require some hassles.

(I'm also open to still adding a new `_internal_inline` header too, if you feel that's the best / cleanest solution).

guangli-dai replied:
I have tested multiple things and still go ahead with keeping `prof_sample_aligned` in `jemalloc_internal_inlines_c.h`. Although it's not ideal, I think it's still acceptable and relatively easy to trace. Besides, the "interdependency" I used before is not accurate. I think the problem here is that the include of different inline headers is strictly ordered [here](https://github.com/jemalloc/jemalloc/blob/dev/include/jemalloc/internal/jemalloc_internal_includes.h#L81). Inline headers included later, i.e., prof_inlines.h, are safe to include other inline headers listed before them, i.e., jemalloc_internal_inlines_c.h. Changing the order to keep `prof_sample_aligned` where it was is not worthy. I think the current modification is the cleanest I can find.

interwq commented on src/jemalloc.c:
```diff
@@ -3040,142 +2987,11 @@ free_default(void *ptr) {
 	}
 }
 
-JEMALLOC_ALWAYS_INLINE bool
```
I assume these are all moved around, w/o any real changes made. Is that the case?

guangli-dai replied:
Yes, no changes made to the codes
